src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Helpers.cs

```diff
@@ -468,5 +468,114 @@ internal static void ThrowIfPasswordContainsNullCharacter(string? password)
+                hashAlgorithmOid,
+                CreateStackTuple(state, callback),
+                static (message, myState) =>
+                {
+                    myState.Item2(message, myState.Item1);
+                    return true;
+                });
+        }
+
+        internal static TResult SlhDsaPreHash<TState, TResult>(
+            ReadOnlySpan<byte> hash,
+            ReadOnlySpan<byte> context,
+            ReadOnlySpan<char> hashAlgorithmOid,
+            TState state,
+            SlhDsaPreHashFuncCallback<byte, TState, TResult> callback)
+            where TState : allows ref struct
+        {
+            // The OIDs for the algorithms above have max length 11. We'll just round up for a conservative initial estimate.
+            const int MaxEncodedOidLengthForCommonHashAlgorithms = 16;
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER, MaxEncodedOidLengthForCommonHashAlgorithms);
+            writer.WriteObjectIdentifier(hashAlgorithmOid);
+
+            int encodedOidLength = writer.GetEncodedLength();
+            int messageLength =
+                1 +                 // Pre-hash encoding flag
+                1 +                 // Context length
+                context.Length +    // Context
+                encodedOidLength +  // OID
+                hash.Length;        // Hash
+
+            const int MessageLengthUpperBoundEstimate = 1 + 1 + 255 + 11 + (512 / 8);
+            Debug.Assert(
+                messageLength <= MessageLengthUpperBoundEstimate,
+                $"Message length {messageLength} exceeds the upper bound of {MessageLengthUpperBoundEstimate}.");
+
+            // Unknown hash algorithms might overshoot the estimated upper bound so we will rent as a backup.
+            byte[]? rented = null;
+            Span<byte> message =
+                messageLength > MessageLengthUpperBoundEstimate
+                    ? (rented = CryptoPool.Rent(messageLength))
+                    : stackalloc byte[MessageLengthUpperBoundEstimate];
+
+            try
+            {
+                // Pre-hash encoding flag
+                message[0] = 0x01;
+
+                // Context length
+                message[1] = checked((byte)context.Length);
+
+                // Context
+                context.CopyTo(message.Slice(2));
+
+                // OID
+                writer.Encode(
+                    message.Slice(2 + context.Length),
+                    (dest, encoded) => encoded.CopyTo(dest));
+
+                // Hash
+                hash.CopyTo(message.Slice(2 + context.Length + encodedOidLength));
+
+                return callback(message.Slice(0, messageLength), state);
+            }
+            finally
+            {
+                if (rented != null)
+                {
+                    CryptoPool.Return(rented);
+                }
+            }
+        }
+
+        internal static StackTuple<T1, T2> CreateStackTuple<T1, T2>(T1 item1, T2 item2)
```

Comment:
could these two items be passed as two args directly in the callback?

Reply:
It's possible, but because I'm calling the Func overload of `SlhDsaPreHash` from the Action overload, the new state needs to contain the previous state and the original callback (to avoid a closure). I can instead implement the Action overload inline without calling into the Func overload. Or just pass in a dummy Func and get rid of the overload: `static () => { Foo(); return 42; }`. I think the current approach is fine though - in most scenarios we would just use a tuple to bundle the state, but since we have to bundle ROS/Span, we need this new ref struct tuple instead.